src/identify/libp2p_identify.erl

```diff
@@ -1,79 +1,72 @@
 -module(libp2p_identify).
 
+-include("pb/libp2p_peer_pb.hrl").
 -include("pb/libp2p_identify_pb.hrl").
 
--opaque identify() :: #libp2p_identify_pb{}.
-
--export_type([identify/0]).
--export([spawn_identify/3, new/4, new/5, encode/1, decode/1,
-         address/1, protocol_version/1, listen_addrs/1, observed_maddr/1, observed_addr/1,
-         protocols/1, agent_version/1]).
-
--define(VERSION, "identify/1.0.0").
-
--spec spawn_identify(Session::pid(), Handler::pid(), UserData::any()) -> pid().
-spawn_identify(Session, Handler, HandlerArgs) ->
-    spawn(fun() ->
-                  libp2p_session:dial_framed_stream(?VERSION, Session,
-                                                    libp2p_stream_identify, [Session, Handler, HandlerArgs])
-          end).
-
--spec new(libp2p_crypto:address(), [string()], string(), [string()])
-         -> libp2p_identify_pb:libp2p_identify_pb().
-new(Address, ListenAddrs, ObservedAddr, Protocols) ->
-    Version = case lists:keyfind(libp2p, 1, application:loaded_applications()) of
-                  {_, _, V} -> V;
-                  false -> "unknown"
-              end,
-    AgentVersion = application:get_env(libp2p, agent_version,
-                                       list_to_binary(["erlang-libp2p/", Version])),
-    new(Address,
-        lists:map(fun multiaddr:new/1, ListenAddrs),
-        multiaddr:new(ObservedAddr),
-        Protocols, AgentVersion).
-
--spec new(libp2p_crypto:address(), [multiaddr:multiaddr()], multiaddr:multiaddr(), [string()], string()) -> identify().
-new(Address, ListenAddrs, ObservedAddr, Protocols, AgentVersion) ->
-    #libp2p_identify_pb{publicKey=Address,
-                        protocol_version=?VERSION,
-                        listen_addrs=ListenAddrs,
-                        observed_addr=ObservedAddr,
-                        protocols=Protocols,
-                        agent_version=AgentVersion
-                       }.
-
--spec protocol_version(identify()) -> string().
-protocol_version(#libp2p_identify_pb{protocol_version=Version}) ->
-     Version.
-
+-type identify() :: #libp2p_signed_identify_pb{}.
+-type identify_map() :: #{ peer => libp2p_peer:peer(),
+                           observed_addr => string(),
+                           nonce => binary()
+                         }.
+-export_type([identify/0, identify_map/0]).
+-export([from_map/2, encode/1, decode/1, verify/1,
+         address/1, peer/1, observed_maddr/1, observed_addr/1, nonce/1]).
+
+-spec from_map(identify_map(), libp2p_crypto:sig_fun()) -> identify().
+from_map(Map, SigFun) ->
+    G = fun(K) ->
+                case maps:get(K, Map, no_entry) of
```

Comment:
maps:get/2 will throw an exception for you, you don't need this wrapper fun...

Reply:
Right but it will  throw a badkey instead of a badarg. I was trying to indicate that you're passing in a bad map. Is that going too much against the grain?

Reply:
It's going against the grain.. simplifying